### PR TITLE
Factuality evaluation: Refactoring, added tests, fixed non-factual entity counting bug & skipped entity counting bug

### PR DESCRIPTION
**Pull request by anton164**
This PR is refactors the evaluation script. The main goal was to add tests & make them pass, then do a bigger refactoring which simplifies the core of the evaluation code by relying on the tests.

**Commit by commit**

1. refactor: moved evaluation methods to src.evaluation.factuality
2. added tests for evaluation of oracle and classifier, note how the tests correspond with these tables:
Screenshot 2022-06-02 at 12 09 26
Screenshot 2022-06-02 at 16 48 34

3. fixed bug in entity labels (non factual hallucinations) being counted even though summary was skipped. Note that this reduces the number of non-factual entities in the classifier as well as the oracle. Tests are now green.

### Update
Added more commits for additional test coverage & completed the refactoring. See comments & commits below.

### CODE REVIEW

**Comment by anton164** [comment]
Added more tests. This should cover all code-paths for evaluation - gold, fbs w/ classifier, fbs w/ oracle and a another model (the baseline).

A test for the corrector / other models shouldn't be needed since it follows the exact same path as the baseline.

**Comment by anton164** [comment]
Added a commit which refactors evaluation. At large evaluation results remain the same, however the entity counts were reduced slightly for FBS oracle & classifier since we were overcounting entities in cases where the summary was skipped.

File: results/debug-oracle.json [hunk 1]
```diff
@@ -971,18 +974,18 @@
...
           "generation_metadata": {
-            "score": -0.5369529128074646,
+            "score": -0.53695148229599,
             "dropped_seqs": [
               "</s>Sam Baldock scored the only goal of the game as Brighton beat Wolves at the Amex Stadium",
               "</s>Sam Baldock scored the only goal of the game as Brighton beat Wolves to move up to third",
               "</s>Sam Baldock's first-half header was enough to give Brighton victory over Wolves at the Amex Stadium"
             ],
-            "n_words_checked": 96
+            "n_words_checked": 81
```
Comment (dleve123) [inline results/debug-oracle.json:988]: why would we expect any diff here? Was thinking the NER in https://github.com/danton-nlp/factual-beam-search/pull/41 - but that's a separate Git history.

**Comment by dleve123** [comment]
> added tests for evaluation of oracle and classifier, note how the tests correspond with these tables:

Just to be clear, the screenshot of the table in the PR reflect not-yet-corrected metrics, for instance, [this diff](https://github.com/danton-nlp/factual-beam-search/pull/40/commits/b2a8583dd1502612e69c7f59da9884524550d5e6#diff-ebd150c40a4f7ef0dff37c0df240c1e765d66e0b3e5514b026d017b76cacde7fR67-R71) is different than the table, but that's because you fixed the entity counting bug in https://github.com/danton-nlp/factual-beam-search/pull/40/commits/43ff2d7dd051c320a31ec593b2dd58ec254d9581

**Comment by anton164** [comment]
> > added tests for evaluation of oracle and classifier, note how the tests correspond with these tables:
> 
> Just to be clear, the screenshot of the table in the PR reflect not-yet-corrected metrics, for instance, [this diff](https://github.com/danton-nlp/factual-beam-search/pull/40/commits/b2a8583dd1502612e69c7f59da9884524550d5e6#diff-ebd150c40a4f7ef0dff37c0df240c1e765d66e0b3e5514b026d017b76cacde7fR67-R71) is different than the table, but that's because you fixed the entity counting bug in [43ff2d7](https://github.com/danton-nlp/factual-beam-search/commit/43ff2d7dd051c320a31ec593b2dd58ec254d9581)

Correct. In the first commit the tests corresponded to the table, but in subsequent commits with fixes some metrics changed.

**Review by dleve123 — COMMENTED** [review]
Comments for the PR for the first 5 commits (through https://github.com/danton-nlp/factual-beam-search/pull/40/commits/b2a8583dd1502612e69c7f59da9884524550d5e6) – reviewing the following commits presently as well.

File: tests/test_evaluation.py [hunk 2]
```diff
@@ -1,6 +1,126 @@
...
+TEST_SIZE = 100
 
-def test_evaluate_factuality():
-    pass
\ No newline at end of file
+
+def load_data(results_path: str):
+    sums_by_id, sum_ents_by_id = load_summaries_from_logs(results_path)
+    gold_sums, gold_metadata = get_gold_data()
+    baseline_metadata = get_summary_metrics("xsum", "facebook-bart-large-xsum")
```
[the remark this answers is not in the document]
Comment (anton164) [reply tests/test_evaluation.py:28]: Yes, `metrics` is a misnomer left over from sumtool. In a future refactoring it'd be great to remove the sumtool dependency and rename these methods locally.

Ideally - I think sumtool should be replaced with a wrapper on top of huggingface datasets.

File: results/debug-classifier-knnv1.json [hunk 3]
```diff
@@ -28,7 +31,6 @@
               "start": 0,
               "end": 8,
               "in_source": true,
-              "predicted_label": "Factual Hallucination",
```
[the remark this answers is not in the document]
Comment (anton164) [reply results/debug-classifier-knnv1.json:31]: Yes, this will be fixed in https://github.com/danton-nlp/factual-beam-search/pull/43

File: results/debug-oracle.json (hunk 1 above)
Comment (anton164) [reply results/debug-oracle.json:988]: I think the issue is that we didn't re-generate `debug-oracle.json` consistently at the end of the deadline push - we were testing & debugging directly with 
the final test-extrinsic set. 

This means that this diff doesn't only capture changes introduced in this PR, but let me rerun debug on main to verify

File: tests/test_evaluation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dleve123) [reply tests/test_evaluation.py:28]: > Ideally - I think sumtool should be replaced with a wrapper on top of huggingface datasets.

Agreed! Just created an issue for it! https://github.com/danton-nlp/factual-beam-search/issues/44

File: results/debug-oracle.json (hunk 1 above)
Comment (dleve123) [reply results/debug-oracle.json:988]: Gotcha, yeah, the debug dset doesn't serve any important purpose anymore either, so all good regardless methinks

Comment (anton164) [reply results/debug-oracle.json]: Reran on main & rebased. These stats dont change anymore

File: src/data_utils.py [hunk 4]
```diff
@@ -2,6 +2,10 @@
 import json
 from datasets import load_dataset
 import random
+from sumtool.storage import get_summary_metrics, get_summaries
+
+SUMTOOL_DATASET = "xsum"
```
[the remark this answers is not in the document]
Comment (anton164) [reply src/data_utils.py:7]: agree, good call

**Comment by anton164** [comment]
@dleve123 good to merge?

File: src/evaluation/factuality.py [hunk 5]
```diff
@@ -97,82 +162,61 @@ def evaluate_factuality(
...
+            summary_eval["count_entity_label"][ANNOTATION_LABELS["Unknown"]] > 0
+        )
+        count_extrinsic = 0
 
-        if summary == SUMMARY_FAILED_GENERATION:
+        # Skip logic
+        is_skipped = False
+        if summary_eval["failed"]:
+            is_skipped = True
+        elif is_fbs and is_oracle and count_non_factual_extrinsic:
```
Comment (dleve123) [inline src/evaluation/factuality.py:198]: Is it possible for FBS in oracle-model to have a more similar API to non-oracle FBS: aka, could it return `SUMMARY_FAILED_GENERATION` as well?

File: src/evaluation/factuality.py [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+                    ] += 1
+            else:
+                summary_eval["count_entity_label"][ent["label"]] += 1
+
+                # Detect non-factual predictions from FBS classifier
+                if (
+                    "predicted_label" in ent
+                    and ent["predicted_label"] == ANNOTATION_LABELS["Non-factual"]
+                ):
+                    summary_eval["has_predicted_non_factual"] = True
```
[the remark this answers is not in the document]
Comment (dleve123) [reply src/evaluation/factuality.py:117]: This seems like the "made mistake case" -- is that right? Can you confirm / provide an example?

File: src/evaluation/factuality.py [hunk 7]
```diff
@@ -187,46 +231,59 @@ def evaluate_factuality(
             print()
             print_counter += 1
 
+        # Increment aggregated results
+        if summary_eval["failed"]:
+            agg_results["failed"] += 1
+
+        # Only increment counts if summary is NOT skipped!
```
Comment (dleve123) [inline src/evaluation/factuality.py:238]: 🙌 – feel like the lack of this (consolidated) conditional was the source of many of the eval bugs, rights?

File: src/evaluation/factuality.py [hunk 8]
```diff
@@ -97,82 +162,61 @@ def evaluate_factuality(
...
-                if ent["in_source"]:
-                    counters[ANNOTATION_LABELS["Factual"]] += 1
-                else:
-                    counters[ANNOTATION_LABELS["Non-hallucinated"]] += 1
-                counters["entities"] += 1
-                continue
-
-            # SKIP LOGIC for FBS
-            if is_fbs:
-                is_oracle = "predicted_label" not in ent
```
Comment (dleve123) [inline src/evaluation/factuality.py]: was `is_oracle` always a misnomer? Why the change?

File: src/evaluation/factuality.py [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+        offset += len(inserted_label)
+    return marked_summary
+
+
+def evaluate_summary(
+    summary: str,
+    source: str,
+    reference: str,
+    labeled_entities: List[MarkedEntity],
+    is_fbs,
```
Comment (dleve123) [inline src/evaluation/factuality.py:72]:
```suggestion
```

seems like arg is unused

File: src/evaluation/factuality.py [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+    is_oracle,
```
Comment (dleve123) [inline src/evaluation/factuality.py:73]:
```suggestion
```

seems like arg is unused

File: src/evaluation/factuality.py [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+        before = marked_summary[:position]
+        after = marked_summary[position:]
+        marked_summary = before + inserted_label + after
+        offset += len(inserted_label)
+    return marked_summary
+
+
+def evaluate_summary(
+    summary: str,
+    source: str,
```
Comment (dleve123) [inline src/evaluation/factuality.py:69]:
```suggestion
    source: str,
```

seems like arg is unused

**Review by dleve123 — CHANGES_REQUESTED** [review]
Generally looks great -- requesting changes because I think there's a bug in `evaluate_summay` in the `is_gold` case -- which I think also exists on `main`, but I very well could be misreading the code

File: src/evaluation/factuality.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (anton164) [reply src/evaluation/factuality.py:117]: No, this is not the "made mistake case", it's for detecting skips  - when the FBS classifier predicts non factual irrespective of the underlying label. 

Detecting "made mistake" is left to a future PR.

File: src/evaluation/factuality.py (hunk 5 above)
Comment (anton164) [reply src/evaluation/factuality.py:198]: Hmm, they both have the same API - `SUMMARY_FAILED_GENERATION` means that the summary generation failed, i.e. no valid beams were left, for both FBS and oracle.

The case you're looking at here is when a summary was generated, but an entity was non-factual. This indicates a skip, but the summary is still returned from the model (it's not the same as `SUMMARY_FAILED_GENERATION`).

File: src/evaluation/factuality.py (hunk 7 above)
Comment (anton164) [reply src/evaluation/factuality.py:238]: Oh yes!

File: src/evaluation/factuality.py [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+        summary_eval["rougeL"] = rouge_scores["rougeL"]["f1"]
+
+        for ent in labeled_entities:
+            summary_eval["count_entity_total"] += 1
+
+            # Increment entity label count
+            if is_gold:
+                if ent["in_source"]:
+                    summary_eval["count_entity_label"][
+                        ANNOTATION_LABELS["Factual"]
```
[the remark this answers is not in the document]
Comment (anton164) [reply src/evaluation/factuality.py:103]: Good catch! Also on main. Didn't have any consequences for results, but obviously worth fixing

File: src/evaluation/factuality.py (hunk 11 above)
Comment (anton164) [reply src/evaluation/factuality.py:69]: Leaving this since access to the source could likely be relevant for summary evaluation

File: src/evaluation/factuality.py [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+        "skipped": False,
+        "failed": False,
+        "count_entity_total": 0,
+        "count_entity_extrinsic": 0,
+        "count_entity_label": defaultdict(lambda: 0),
+        "has_predicted_non_factual": False,
+        "rouge1": None,
+        "rouge2": None,
+        "rougeL": None,
+    }
```
[the remark this answers is not in the document]
Comment (anton164) [reply src/evaluation/factuality.py]: Added - altho I'm still not convinced that python typing adds value in its current form. In this project it has been marginally helpful for reading code & thinking about code in terms of types, but is hardly able to catch any bugs & adds maintenance overhead.

File: tests/test_evaluation.py [hunk 14]
```diff
@@ -4,14 +4,26 @@
...
+    sums_by_id, sum_ents_by_id = (
+        load_summaries_from_logs(results_path)
+        if "results" in results_path
+        # load from sumtool
+        else (
+            {
+                sum_id: x["summary"]
+                for sum_id, x in get_summaries("xsum", results_path).items()
+            },
+            {},
```
[the remark this answers is not in the document]
Comment (anton164) [reply tests/test_evaluation.py:24]: Correct, the entities are detected on the fly for those summaries.

File: tests/test_evaluation.py [hunk 15]
```diff
@@ -4,14 +4,26 @@
...
+        )
+    )
```
[the remark this answers is not in the document]
Comment (anton164) [reply tests/test_evaluation.py]: agree, done

File: src/evaluation/factuality.py (hunk 8 above)
Comment (anton164) [reply src/evaluation/factuality.py]: The logic of this method has been completely refactored so hard to explain why this line changed - unsure what you mean by misnomer.

The `is_oracle` flag was correct, but has been moved to the function args and is no longer derived from "predicted_label".

**Comment by anton164** [comment]
@dleve123 comments addressed! Thanks for the thorough review

File: src/evaluation/factuality.py (hunk 5 above)
Comment (dleve123) [reply src/evaluation/factuality.py:206]: makes sense!

**Review by dleve123 — APPROVED** [review]
